trax/models/research/funnel_transformer.py

```diff
@@ -0,0 +1,269 @@
+      tl.LayerNorm(),                   # funnel_activations, mask
+      tl.Parallel(
+          None,
+          tl.Fn('mask_max_pool',
+                _InternalMaxPool),
+      ),                                # funnel_activations, mask'
+      feed_forward
+  )
+
+
+def FunnelTransformerEncoder(vocab_size,
+                             n_classes=10,
+                             d_model=512,
+                             d_ff=2048,
+                             encoder_segment_lengths=(2, 2, 2),
+                             n_heads=8,
+                             max_len=2048,
+                             dropout=0.1,
+                             dropout_shared_axes=None,
+                             mode='train',
+                             ff_activation=tl.Relu,
+                             pool_layer=tl.AvgPool,
+                             pool_size=(2,),
+                             strides=(2,),
+                             separate_cls=True):
+  """Returns a Funnel Encoder.
+  """
+  assert encoder_segment_lengths
+
+  positional_encoder = [
+      tl.Embedding(vocab_size, d_model),
+      tl.Dropout(rate=dropout, shared_axes=dropout_shared_axes, mode=mode),
+      tl.PositionalEncoding(max_len=max_len)]
+
+  encoder_blocks = []
+  n_encoder_segments = len(encoder_segment_lengths)
+
+  for i in range(n_encoder_segments):
+    # Building i'th segment
+    for _ in range(encoder_segment_lengths[i]):
+      # Create segment_size encoder blocks
+      encoder_blocks.append(
+          _EncoderBlock(d_model, d_ff, n_heads, dropout, dropout_shared_axes,
+                        mode, ff_activation))
+
+    # If not last segment, add funnel block
+    if i != n_encoder_segments - 1:
+      encoder_blocks.append(_FunnelBlock(d_model, d_ff, n_heads, dropout,
+                                         dropout_shared_axes, mode,
+                                         ff_activation, pool_layer, pool_size,
+                                         strides, separate_cls))
+
+  # Assemble and return the model.
+  return tl.Serial(                               # toks
+      # Encode.
+      tl.Branch(
+          positional_encoder, tl.PaddingMask()),  # vecs masks
+      encoder_blocks,                             # vecs masks
+      tl.Select([0], n_in=2),                     # vecs
+      tl.LayerNorm(),                             # vecs
+
+      # Map to output categories.
+      tl.Mean(axis=1),                            # vecs
+      tl.Dense(n_classes),                        # vecs
+      tl.LogSoftmax(),                            # vecs
+  )
+
+
+def _FunnelResidualBlock(d_model, d_ff, n_heads,
+                         dropout, dropout_shared_axes, mode, ff_activation,
+                         pool_layer, pool_size, strides):
+  feed_forward = _FeedForwardBlock(
+      d_model, d_ff, dropout, dropout_shared_axes, mode, ff_activation)
+
+  dropout_ = tl.Dropout(
```

Comment:
Why do variables have "_" suffix here?

Reply:
Fixed - `dropout_` layer variable name was taken from the original TransformerEncoder to avoid shadowing `dropout` rate argument, replaced by a more meaningful `hidden_dropout`.